src/react/Reconciler.js

```diff
@@ -0,0 +1,293 @@
+    return {
+      result,
+      childContext,
+      commitDidMountPhase,
+    };
+  }
+  _renderOneLevel(
+    componentType: ECMAScriptSourceFunctionValue,
+    props: ObjectValue | AbstractValue,
+    context: ObjectValue | AbstractValue,
+    branchStatus: BranchStatusEnum
+  ) {
+    if (valueIsClassComponent(this.realm, componentType)) {
+      // for now we don't support class components, so we throw a ExpectedBailOut
+      throw new ExpectedBailOut("class components not yet supported");
+    } else {
+      invariant(componentType.$Call, "Expected componentType to be a FunctionValue with $Call method");
+      let value = componentType.$Call(this.realm.intrinsics.undefined, [props, context]);
+      return { value, commitDidMountPhase: null, childContext: context };
+    }
+  }
+  _applyBranchedLogic(value: ObjectValue) {
+    // we need to apply a key when we're branched
+    let currentKeyValue = Get(this.realm, value, "key") || this.realm.intrinsics.null;
+    let uniqueKey = getUniqueReactElementKey("", this.reactSerializerState.usedReactElementKeys);
+    let newKeyValue = GetValue(this.realm, this.realm.$GlobalEnv.evaluate(t.stringLiteral(uniqueKey), false));
+    if (currentKeyValue !== this.realm.intrinsics.null) {
+      newKeyValue = computeBinary(this.realm, "+", currentKeyValue, newKeyValue);
+    }
+    value.$Set("key", newKeyValue, value);
+    return value;
+  }
+  _updateBranchStatus(branchStatus: BranchStatusEnum): BranchStatusEnum {
+    if (branchStatus === BranchStatus.NEW_BRANCH) {
+      branchStatus = BranchStatus.BRANCH;
+    }
+    return branchStatus;
+  }
+  _resolveDeeply(value: Value, context: ObjectValue | AbstractValue, branchStatus: BranchStatusEnum) {
+    if (
+      value instanceof StringValue ||
+      value instanceof NumberValue ||
+      value instanceof BooleanValue ||
+      value instanceof NullValue ||
+      value instanceof UndefinedValue
+    ) {
+      // terminal values
+      return value;
+    } else if (value instanceof AbstractValue) {
+      for (let i = 0; i < value.args.length; i++) {
+        value.args[i] = this._resolveDeeply(value.args[i], context, BranchStatus.NEW_BRANCH);
+      }
+      return value;
+    }
+    if (value instanceof ArrayValue) {
+      this._resolveFragment(value, context, branchStatus);
+      return value;
+    }
+    if (value instanceof ObjectValue && isReactElement(value)) {
+      let typeValue = Get(this.realm, value, "type");
+      let propsValue = Get(this.realm, value, "props");
+      let refValue = Get(this.realm, value, "ref");
+      if (typeValue instanceof StringValue) {
+        // terminal host component. Start evaluating its children.
+        if (propsValue instanceof ObjectValue) {
+          let childrenProperty = propsValue.properties.get("children");
+          if (childrenProperty) {
+            let childrenPropertyDescriptor = childrenProperty.descriptor;
+            invariant(childrenPropertyDescriptor, "");
```

Comment:
Why is this invariant true? I think the descriptor can be absent for deleted properties.